glean/src/core/metrics/types/string_list.ts

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import type { CommonMetricData } from "../index.js";
+import { MetricType } from "../index.js";
+import { Context } from "../../context.js";
+import { Metric } from "../metric.js";
+import { isString, truncateStringAtBoundaryWithError } from "../../utils.js";
+import type { JSONValue } from "../../utils.js";
+import { ErrorType } from "../../error/error_type.js";
+
+export const MAX_LIST_LENGTH = 20;
+export const MAX_STRING_LENGTH = 50;
+
+export class StringListMetric extends Metric<string[], string[]> {
+  constructor(v: unknown) {
+    super(v);
+  }
+
+  validate(v: unknown): v is string[] {
+    if (!Array.isArray(v)) {
+      return false;
+    }
+
+    if (v.length > MAX_LIST_LENGTH) {
+      return false;
+    }
+    for (const s of v) {
+      if (!isString(s) || s.length > MAX_STRING_LENGTH) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  payload(): string[] {
+    return this._inner;
+  }
+}
+
+/**
+ * A string list metric.
+ *
+ * This allows appending a string value with arbitrary content to a list.
+ * The list is length-limited to `MAX_LIST_LENGTH`.
+ * Strings are length-limited to `MAX_STRING_LENGTH` bytes.
+ */
+class StringListMetricType extends MetricType {
+
+  constructor(meta: CommonMetricData) {
+    super("string_list", meta);
+  }
+
+  /**
+   * Sets to the specified string list value.
+   *
+   * # Note
+   *
+   * Truncates the list if it is longer than `MAX_LIST_LENGTH` and logs an error.
+   *
+   * Truncates the value if it is longer than `MAX_STRING_LENGTH` bytes
+   * and logs an error.
+   *
+   * @param value the value to set.
+   */
+  set(value: string[]): void {
+    if (!this.shouldRecord(Context.uploadEnabled)) {
+      return;
+    }
+
+    if (value.length > MAX_LIST_LENGTH) {
+      Context.errorManager.record(
+        this,
+        ErrorType.InvalidValue,
+        `String list length of ${value.length} exceeds maximum of ${MAX_LIST_LENGTH}.`
+      );
+    }
+
+    const truncatedList: string[] = [];
+    const newLength = Math.min(value.length, MAX_LIST_LENGTH);
+    for (let i = 0; i < newLength; ++i) {
+      const truncatedString = truncateStringAtBoundaryWithError(this, value[i], MAX_STRING_LENGTH);
+      truncatedString.then((v) => {
+        truncatedList.push(v);
+      });
+    }
+    const metric = new StringListMetric(truncatedList);
+    Context.metricsDatabase.record(this, metric);
```

Comment:
The whole body needs to be wrapped in a `dispatcher.launch` call, like so:

```suggestion
Context.dispatcher.launch(async () => {
    if (!this.shouldRecord(Context.uploadEnabled)) {
      return;
    }

    if (value.length > MAX_LIST_LENGTH) {
      Context.errorManager.record(
        this,
        ErrorType.InvalidValue,
        `String list length of ${value.length} exceeds maximum of ${MAX_LIST_LENGTH}.`
      );
    }

    const truncatedList: string[] = [];
    const newLength = Math.min(value.length, MAX_LIST_LENGTH);
    for (let i = 0; i < newLength; ++i) {
      const truncatedString = truncateStringAtBoundaryWithError(this, value[i], MAX_STRING_LENGTH);
      truncatedString.then((v) => {
        truncatedList.push(v);
      });
    }
    const metric = new StringListMetric(truncatedList);
    Context.metricsDatabase.record(this, metric);
 });
```

Reply:
Hi @brizental , I just want to make sure that you still want me to merge `_private_addUndispatched` and `add`, is that correct? Or does it make more sense to keep `_private_addUndispatched`?

Reply:
On a second look, I keep my initial opinion that it is not necessary to have the undispatched methods.